### PR TITLE
Welcome page cleanup

### PR DESCRIPTION
**Pull request by jryans**
Fixed up a few edge cases with the welcome page, such as:

* Can't go to room from room dir in welcome area (https://github.com/vector-im/riot-web/issues/8480)
* Can't go to direct room links from app start (https://github.com/vector-im/riot-web/issues/8476)
* Can't join a room when logged out from room view

Used by https://github.com/vector-im/riot-web/pull/8525

### CODE REVIEW

File: src/components/structures/RoomView.js [hunk 1]
```diff
@@ -877,13 +877,12 @@ module.exports = React.createClass({
...
             // Join this room once the user has registered and logged in
-            const signUrl = this.props.thirdPartyInvite ?
-                this.props.thirdPartyInvite.inviteSignUrl : undefined;
+            // (If we failed to peek, we may not have a valid room object.)
             dis.dispatch({
                 action: 'do_after_sync_prepared',
                 deferred_action: {
                     action: 'view_room',
-                    room_id: this.state.room.roomId,
+                    room_id: this.state.room ? this.state.room.roomId : this.state.roomId,
```
Comment (bwindels) [inline src/components/structures/RoomView.js]: Maybe we should always use `state.roomId`?

Comment (jryans) [reply src/components/structures/RoomView.js]: Well, I wasn't sure if we keep it up to date... In `_onRoomViewStoreUpdate`, we set the new state, and it specifically mentions `this.state.roomId` may be null if it we have a room alias:

https://github.com/matrix-org/matrix-react-sdk/blob/9d3ba2b3d9a9fcc0f8ee1fc222c9da9d26ce3d25/src/components/structures/RoomView.js#L273-L283

I didn't notice any code paths that might later go back and update `this.state.roomId` after resolving an alias, so it seemed safer to check both.

@bwindels I haven't looked at this area much, so what would you recommend?

Comment (jryans) [reply src/components/structures/RoomView.js]: For now, I'll add a comment about this.

Comment (bwindels) [reply src/components/structures/RoomView.js]: Yes, sounds good, I forgot about room alias resolving.

**Pull request merged**